### PR TITLE
[BEAM-10289] Dynamic splitting implementation.

### PR DESCRIPTION
**Pull request by youngoli**
Adds dynamic splitting support and tests for it.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by youngoli** [comment]
CC: @lostluck 

Doesn't need a review yet, but if you want to give it a look, have at it.

File: sdks/go/pkg/beam/core/runtime/exec/datasource.go [hunk 1]
```diff
@@ -302,31 +302,67 @@ func (n *DataSource) Split(splits []int64, frac float64, bufSize int64) (int64,
...
+		if err != nil {
+			return SplitResult{}, err
+		}
+
+		if p != nil && r != nil { // Successful split.
+			pEnc, err := encodeElm(p, n.Coder)
+			if err != nil {
+				return SplitResult{}, err
+			}
+			rEnc, err := encodeElm(r, n.Coder)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/exec/datasource.go:346]: Constructing the encoders every split request is expensive, so we're best off creating those ahead of time, even if they aren't necessarily going to be used every time. Unlike on the main ProcessElement path, we can't just keep the encoders around in a function scope, so we'll have to just add it to the Source. But we can just construct them on demand at least, which is much cheaper than repeatedly constructing them.

At the very least, ideally we reuse them between the Primary and the Residual. It's likely just noise for the repeated constructions for Successful splits, but we might as well do half as much work.

In short, we should wait until we see this coder construction in a CPU profile before the heavier caching is applied, but we definitely should avoid the double work here.

File: sdks/go/pkg/beam/core/runtime/exec/translate.go [hunk 2]
```diff
@@ -418,7 +418,8 @@ func (b *builder) makeLink(from string, id linkID) (Node, error) {
 					}
 					u = n
 					if urn == urnProcessSizedElementsAndRestrictions {
-						u = &ProcessSizedElementsAndRestrictions{PDo: n}
+						transform.GetInputs()
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/exec/translate.go:421]: Copy Pasta?

File: sdks/go/pkg/beam/core/runtime/exec/sdf.go [hunk 3]
```diff
@@ -325,6 +337,63 @@ func (n *ProcessSizedElementsAndRestrictions) String() string {
...
+		err := errors.New("Restriction tracker missing.")
+		return nil, nil, errors.WithContext(err, "Attempting split in ProcessSizedElementsAndRestrictions")
+	}
+	p, r, err := n.rt.TrySplit(f)
+	if err != nil {
+		return nil, nil, errors.WithContext(err, "Attempting split in ProcessSizedElementsAndRestrictions")
+	}
+
+	var pfv, rfv *FullValue
+	if r != nil { // If r is nil then the split failed/returned an empty residual.
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/exec/sdf.go:358]: Here's another opportunity to invert the statement for clarity. 

eg.

if r == nil {
  return nil, nil, nil
}

rest of the function.

File: sdks/go/pkg/beam/core/runtime/exec/translate.go (hunk 2 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/exec/translate.go:421]: Something like that. Definitely not supposed to be there. Removed.

File: sdks/go/pkg/beam/core/runtime/exec/datasource.go [hunk 4]
```diff
@@ -302,31 +302,67 @@ func (n *DataSource) Split(splits []int64, frac float64, bufSize int64) (int64,
...
 		bufSize = n.splitIdx
 	}
-	s, _, err := splitHelper(n.index, bufSize, currProg, splits, frac, false)
+	s, f, err := splitHelper(n.index, bufSize, currProg, splits, frac, su != nil)
 	if err != nil {
-		n.mu.Unlock()
-		return 0, err
+		return SplitResult{}, err
+	}
+	if f > 0.0 {
```
[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/exec/datasource.go:334]: Sounds good to me. Done.

File: sdks/go/pkg/beam/core/runtime/exec/datasource.go (hunk 1 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/exec/datasource.go:346]: Good point. Done.

File: sdks/go/pkg/beam/core/runtime/exec/sdf.go (hunk 3 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/exec/sdf.go:358]: Agreed, done.

**Comment by youngoli** [comment]
Run Go PostCommit

**Comment by youngoli** [comment]
Note that this PR has absorbed #12046 so that I can test this change. So if this gets merged in, that PR can be closed. (Alternatively, we can merge that in first, so that this PR is easier to review. Whatever works.)

**Comment by lostluck** [comment]
Consider updating the PR description to match the current state :)

**Comment by youngoli** [comment]
Good idea.

File: sdks/go/pkg/beam/core/runtime/exec/datasource.go [hunk 5]
```diff
@@ -47,20 +47,21 @@ type DataSource struct {
 	splitIdx  int64
 	start     time.Time
 
-	// rt is non-nil if this DataSource feeds directly to a splittable unit,
-	// and receives the current restriction tracker being processed.
-	rt chan sdf.RTracker
+	// su is non-nil if this DataSource feeds directly to a splittable unit,
+	// and receives that splittable unit when it is available for splitting.
+	su chan SplittableUnit
```
[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/exec/datasource.go:52]: Improved this documentation for now, but this is part of concurrency code I might be refactoring to avoid all the blocking, so it might just be gone with the next commit I send.

File: sdks/go/pkg/beam/core/runtime/exec/datasource.go [hunk 6]
```diff
@@ -302,31 +308,76 @@ func (n *DataSource) Split(splits []int64, frac float64, bufSize int64) (int64,
...
+		return SplitResult{PI: s - 1, RI: s}, nil
+	}
+	// Otherwise, perform a sub-element split.
+	fr := f / (1.0 - currProg)
+	p, r, err := su.Split(fr)
+	if err != nil {
+		return SplitResult{}, err
+	}
+
+	if p != nil && r != nil { // Successful split.
```
[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/exec/datasource.go:353]: Fixed.

File: sdks/go/pkg/beam/core/runtime/exec/datasource_test.go [hunk 7]
```diff
@@ -278,10 +279,15 @@ func TestDataSource_Split(t *testing.T) {
 			runOnRoots(ctx, t, p, "StartBundle", func(root Root, ctx context.Context) error { return root.StartBundle(ctx, "1", dc) })
 
 			// SDK never splits on 0, so check that every test.
-			if splitIdx, err := p.Split(SplitPoints{Splits: []int64{0, test.splitIdx}}); err != nil {
+			if splitRes, err := p.Split(SplitPoints{Splits: []int64{0, test.splitIdx}}); err != nil {
 				t.Fatalf("error in Split: %v", err)
-			} else if got, want := splitIdx, test.splitIdx; got != want {
-				t.Fatalf("error in Split: got splitIdx = %v, want %v ", got, want)
+			} else {
```
[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/exec/datasource_test.go:284]: Done.

File: sdks/go/pkg/beam/core/runtime/exec/datasource_test.go [hunk 8]
```diff
@@ -342,10 +348,15 @@ func TestDataSource_Split(t *testing.T) {
 					<-blockedCh
 					// Validate that we do not split on the element we're blocking on index.
 					// The first valid split is at test.splitIdx.
-					if splitIdx, err := source.Split([]int64{0, 1, 2, 3, 4, 5}, -1, 0); err != nil {
+					if splitRes, err := source.Split([]int64{0, 1, 2, 3, 4, 5}, -1, 0); err != nil {
 						t.Errorf("error in Split: %v", err)
-					} else if got, want := splitIdx, test.splitIdx; got != want {
-						t.Errorf("error in Split: got splitIdx = %v, want %v ", got, want)
+					} else {
```
[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/exec/datasource_test.go:353]: This one can't avoid the else block since the error case is `t.Errorf` instead of `t.Fatalf`.

File: sdks/go/pkg/beam/core/runtime/exec/datasource_test.go [hunk 9]
```diff
@@ -412,17 +423,120 @@ func TestDataSource_Split(t *testing.T) {
 
 		// SDK never splits on 0, so check that every test.
 		sp := SplitPoints{Splits: test.splitPts, Frac: test.frac, BufSize: test.bufSize}
-		if splitIdx, err := p.Split(sp); err != nil {
+		if splitRes, err := p.Split(sp); err != nil {
 			t.Fatalf("error in Split: %v", err)
-		} else if got, want := splitIdx, test.splitIdx; got != want {
-			t.Fatalf("error in Split: got splitIdx = %v, want %v ", got, want)
+		} else {
```
[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/exec/datasource_test.go:428]: Done.

File: sdks/go/pkg/beam/core/runtime/exec/dynsplit_test.go [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+type splitResult struct {
+	split SplitResult
+	err   error
+}
+
+// splitPlan is meant to be the goroutine representing the thread handling a
+// split request for the SDF.
+func splitPlan(plan *Plan, result chan splitResult) {
+	// Fake splitting code.
+	//p, r, err := rt.TrySplit(0.5)
```
[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/exec/dynsplit_test.go:322]: Nope, just forgot to delete it.

File: sdks/go/pkg/beam/core/runtime/exec/sdf.go [hunk 11]
```diff
@@ -215,14 +215,25 @@ func (n *SplitAndSizeRestrictions) String() string {
...
+	ctInv   *ctInvoker
+	sizeInv *rsInvoker
+
+	// SU is a buffered channel for indicating when this unit is splittable.
+	// When this unit is processing an element, it sends a SplittableUnit
+	// interface through the channel. That interface can be received on other
+	// threads and used to perform splitting or other related operation.
+	//
+	// Receiving the SplittableUnit prevents the current element from finishing
+	// processing, so he element does not unexpectedly change during a split.
```
[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/exec/sdf.go:229]: Fixed.

**Comment by youngoli** [comment]
Added a new commit addressing the latest comments, and also adding a timeout to the usage of the SplittableUnit channel in DataSource, to avoid the possibility of it blocking for arbitrarily long periods of time.

The timeout I chose was arbitrary, and the idea of choosing a timeout in the first place instead of just a default case is to avoid the subtle edge case where a split occurs right after the current element has incremented in DataSource, but before that element has begun processing in the splittable unit. Without the timeout, those cases will result in avoiding a sub-element split when one was possible if we just waited slightly longer. But maybe I'm just overthinking it, and that case is rare enough that it's not worth worrying about. I'm still pretty undecided on it, and I'd like to hear your opinion @lostluck

**Comment by youngoli** [comment]
Run Go PostCommit
